sw/host/opentitanlib/src/util/present.rs

```diff
@@ -0,0 +1,314 @@
+    0x0c, 0x05, 0x06, 0x0b, 0x09, 0x00, 0x0a, 0x0d, 0x03, 0x0e, 0x0f, 0x08, 0x04, 0x07, 0x01, 0x02,
+];
+
+const S_BOX_INV: [u8; 16] = [
+    0x05, 0x0e, 0x0f, 0x08, 0x0c, 0x01, 0x02, 0x0d, 0x0b, 0x04, 0x06, 0x03, 0x00, 0x07, 0x09, 0x0a,
+];
+
+const P_BOX: [u8; 64] = [
+    0x00, 0x10, 0x20, 0x30, 0x01, 0x11, 0x21, 0x31, 0x02, 0x12, 0x22, 0x32, 0x03, 0x13, 0x23, 0x33,
+    0x04, 0x14, 0x24, 0x34, 0x05, 0x15, 0x25, 0x35, 0x06, 0x16, 0x26, 0x36, 0x07, 0x17, 0x27, 0x37,
+    0x08, 0x18, 0x28, 0x38, 0x09, 0x19, 0x29, 0x39, 0x0a, 0x1a, 0x2a, 0x3a, 0x0b, 0x1b, 0x2b, 0x3b,
+    0x0c, 0x1c, 0x2c, 0x3c, 0x0d, 0x1d, 0x2d, 0x3d, 0x0e, 0x1e, 0x2e, 0x3e, 0x0f, 0x1f, 0x2f, 0x3f,
+];
+
+const P_BOX_INV: [u8; 64] = [
+    0x00, 0x04, 0x08, 0x0c, 0x10, 0x14, 0x18, 0x1c, 0x20, 0x24, 0x28, 0x2c, 0x30, 0x34, 0x38, 0x3c,
+    0x01, 0x05, 0x09, 0x0d, 0x11, 0x15, 0x19, 0x1d, 0x21, 0x25, 0x29, 0x2d, 0x31, 0x35, 0x39, 0x3d,
+    0x02, 0x06, 0x0a, 0x0e, 0x12, 0x16, 0x1a, 0x1e, 0x22, 0x26, 0x2a, 0x2e, 0x32, 0x36, 0x3a, 0x3e,
+    0x03, 0x07, 0x0b, 0x0f, 0x13, 0x17, 0x1b, 0x1f, 0x23, 0x27, 0x2b, 0x2f, 0x33, 0x37, 0x3b, 0x3f,
+];
+
+/// Generate the round_keys for an 80-bit key.
+fn generate_round_keys_80(key: &[u8], rounds: usize) -> Vec<u64> {
+    let mut key = Vec::from(key);
+    let mut round_keys = Vec::new();
+    for i in 1..rounds as u8 + 1 {
+        let mut next_key = key.clone();
+
+        // rawKey[0:64]
+        let mut round_key: u64 = 0;
+        for j in 0..8 {
+            round_key |= (key[7 - j] as u64) << (j * 8);
+        }
+
+        // 1. Shift
+        // rawKey[19:len(rawKey)]+rawKey[0:19]
+        for j in 0..10 {
+            next_key[j] = key[(j + 8) % 10] >> 3 | key[(j + 7) % 10] << 5;
+        }
+
+        // 2. SBox
+        // rawKey[76:80] = S(rawKey[76:80])
+        next_key[0] = S_BOX[(next_key[0] >> 4) as usize] << 4 | next_key[0] & 0x0f;
+
+        // 3. Salt
+        // rawKey[15:20] ^ i
+        next_key[7] ^= i >> 1;
+        next_key[8] ^= i << 7;
+        key = next_key;
+        round_keys.push(round_key);
+    }
+
+    round_keys
+}
+
+/// Generate the round_keys for a 128-bit key.
+fn generate_round_keys_128(key: &[u8], rounds: usize) -> Vec<u64> {
+    let mut key = Vec::from(key);
+    let mut round_keys = Vec::new();
+    for i in 1..rounds as u8 + 1 {
+        let mut next_key = key.clone();
+        let mut round_key: u64 = 0;
+        for j in 0..8 {
+            round_key |= (key[7 - j] as u64) << (j * 8);
+        }
+        for j in 0..16 {
+            next_key[j] = key[(j + 8) % 16] >> 3 | key[(j + 7) % 16] << 5;
+        }
+        next_key[0] =
+            S_BOX[(next_key[0] >> 4) as usize] << 4 | S_BOX[(next_key[0] & 0x0f) as usize];
+        next_key[7] ^= i >> 2;
+        next_key[8] ^= i << 6;
+        key = next_key;
+        round_keys.push(round_key);
+    }
+
+    round_keys
+}
+
+/// SBox funciton for encryption.
+fn s_box_layer(state: u64) -> u64 {
+    let mut output: u64 = 0;
+    for i in 0..16 {
```

Comment:
this looks like it wants to be `(0..16).step(4)`. (or step_by, can't remember what it's called)